File: c/cwe127/001.c
// cwes : ["cwe127", "cwe188"]

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// struct used to force the memory layout
struct frame {
  char secret[32];
  char buffer[32];
};

char *vuln(int index, int size) {
  struct frame f = {
      .secret = "guacamole",
      .buffer = "Lorem ipsum dolor sit amet, con",
  };

  printf("secret : %p\n", f.secret);
  printf("buffer : %p\n", f.buffer);
  printf("bad : %p\n", f.buffer + index);

  if (index + size < 32 && index < 32) {
    char *ret = malloc(size);
    memcpy(ret, f.buffer + index, size);
    return ret;
  }

  return NULL;
}

int main() {
  printf("%s\n", vuln(5, 10));   // fine
  printf("%s\n", vuln(-32, 10)); // vuln
}

File: rust/cwe127/001.rs
const BUFFER_SIZE: usize = 32;

#[repr(C)]
struct Frame {
    secret: [u8; BUFFER_SIZE],
    buffer: [u8; BUFFER_SIZE],
}

fn vuln(index: i32, size: i32) -> Result<String, ()> {
    let f = Frame {
        secret: {
            let tmp = b"guacamole\0";
            let mut buffer: [u8; BUFFER_SIZE] = [b'\0'; BUFFER_SIZE];
            buffer[..tmp.len()].copy_from_slice(tmp);
            buffer
        },
        buffer: {
            let tmp = b"Lorem ipsum dolor sit amet, con\0";
            let mut buffer: [u8; BUFFER_SIZE] = [b' '; BUFFER_SIZE];
            buffer[..tmp.len()].copy_from_slice(tmp);
            buffer
        },
    };

    if index + size < 32 && index < 32 {
        let mut out = Vec::with_capacity(size as usize);

        unsafe {
            // dbg!("here", std::ffi::CStr::from_ptr(f.buffer.as_ptr() as *const i8));
            std::ptr::copy_nonoverlapping(
                f.buffer.as_ptr().offset(index as isize),
                out.as_mut_ptr(),
                size as usize,
            );
        }
        // dbg!(*f.buffer.as_ptr().offset(index as isize));
        // }
        // dbg!(&out);
        return Ok(unsafe {
            std::ffi::CStr::from_ptr(out.as_mut_ptr() as *const i8)
                .to_str()
                .unwrap()
                .to_string()
        });
    }

    return Err(());
}

fn main() {
    dbg!(vuln(5, 10));
    dbg!(vuln(-32, 10));
}
